File: sdk/ruby/test/wire/channel_messages_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class ChannelMessagesWireTest < WireMockTestCase
  def setup
    super

    @client = AgentMailer::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_channel_messages_list_with_wiremock
    test_id = "channel_messages.list.0"

    @client.channel_messages.list(request_options: {
      additional_headers: {
        "X-Test-Id" => "channel_messages.list.0"
      }
    })

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/channel-messages",
      query_params: nil,
      expected: 1
    )
  end

  def test_channel_messages_send__with_wiremock
    test_id = "channel_messages.send_.0"

    @client.channel_messages.send_(
      idempotency_key: "Idempotency-Key",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "channel_messages.send_.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/channel-messages",
      query_params: nil,
      expected: 1
    )
  end
end

File: sdk/rust/tests/channel_messages_test.rs
use agentmailer::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_channel_messages_list_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .channel_messages
        .list(
            &ChannelMessagesListQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/channel-messages", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_channel_messages_send_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .channel_messages
        .send(
            &ChannelMessageSend::Sms {
                data: ChannelMessageSendSms {
                    endpoint_id: "endpointId".to_string(),
                    to: "to".to_string(),
                    consent: ChannelMessageSendSmsConsent {
                        basis: ChannelMessageSendSmsConsentBasis::Explicit,
                        captured_at: DateTime::parse_from_rfc3339("2024-01-15T09:30:00Z").unwrap(),
                        reference: "reference".to_string(),
                    },
                    content: ChannelMessageSendSmsContent {
                        r#type: ChannelMessageSendSmsContentType::Text,
                        body: "body".to_string(),
                    },
                    message_type: None,
                },
            },
            Some(RequestOptions::new().additional_header("Idempotency-Key", "Idempotency-Key")),
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/channel-messages", None, 1)
        .await
        .unwrap();
}
